### DISCUSSION TITLE
Remove ORC, OBR, OBX(s), NTE(s) segments

### DISCUSSION THREAD

**Discussion by Electro666** [nextgenhealthcare/connect#5889] [Q&A] [answered]
pacmano1 assisted to articulate my question, thank-you:

I want to delete an ORC segment and any subsequent segments after that ORC up the next ORC if any OBR after that ORC (and before the
next ORC) has OBR.4.1 = '12345'.

I tried using 3 functions below and I can only delete OBR and OBXs.

// original thread - https://github.com/nextgenhealthcare/connect/discussions/5244

```
// delete javascript array of segments
function deleteSegList(list) {
	list.forEach(seg => {delete seg[0]})
}

// delete OBX segment and consecutive NTE segments
function deleteObservation(obx) {
	var nteList = getSegmentsAfter(msg, obx, 'NTE', true)
	deleteSegList(nteList)
	delete obx[0]
}

// delete OBR segment and all consecutive segments until the next OBR or end of message
function deleteOrderObservation(obr) {
	var allSegsUntilNextObr = getSegmentsAfter(msg, obr, /.*/)
	deleteSegList(allSegsUntilNextObr)
	delete obr[0]
}
```

**Code**
```
for each (var obr in msg.OBR) {
	var obxList = getSegmentsAfter(msg, obr, 'OBX')
if (obr['OBR.4']['OBR.4.1'].toString() == '12345') {
	deleteOrderObservation(obr)
	}
}
```

**Sample msg**
```
MSH|^~\&|TEST|TEST|TEST|TEST|20230817104747||ORU^R01|16967118590||2.3|
PID|1|88|88||
PV1|1||||||||||||||||||||
ORC|RE|22222||||||||||
OBR|1|22222|678|88888^TestEIGHT||
OBX|1|NM|88888^Test||
NTE|1||This test was developed and its performance characteristics
NTE|2||determined by Lab. It has not been cleared or
ORC|RE|22222||||||||||
OBR|1|22222|678|12345^Test12345||
OBX|1|NM|123459999^Test||
NTE|1||This test was developed and its performance characteristics
NTE|2||determined by Lab. It has not been cleared or
ORC|RE|22222||||||||||
OBR|1|22222|678|66666^TestSIX||
OBX|1|NM|66666^Test||
NTE|1||This test was developed and its performance characteristics
NTE|2||determined by Lab. It has not been cleared or
```

**Desired output msg**
```
MSH|^~\&|TEST|TEST|TEST|TEST|20230817104747||ORU^R01|16967118590||2.3|
PID|1|88|88||
PV1|1||||||||||||||||||||
ORC|RE|22222||||||||||
OBR|1|22222|678|88888^TestEIGHT||
OBX|1|NM|88888^Test||
NTE|1||This test was developed and its performance characteristics
NTE|2||determined by Lab. It has not been cleared or
ORC|RE|22222||||||||||
OBR|1|22222|678|66666^TestSIX||
OBX|1|NM|66666^Test||
NTE|1||This test was developed and its performance characteristics
NTE|2||determined by Lab. It has not been cleared or
```

**Comment by Electro666** [comment]
I couldn't get the function to work to remove ORC/OBR/OBX if OBR.4.1 = '12345', so I created logic below that will handle it.  If there's a better way I would love to try it.

```
// first delete OBX
for each (obr in msg.OBR)
	if (obr['OBR.4']['OBR.4.1'].toString() == '12345') {
		for each (obx in getSegmentsAfter(msg,obr,'OBX',false))
			delete msg.children()[obx.childIndex()];
}

// next delete ORC and OBR
for (var i=0; i<msg['ORC'].length(); i++ ) {
	if (msg['OBR'][i]['OBR.4']['OBR.4.1'].toString() == '12345') {
		delete msg['ORC'][i];
		delete msg['OBR'][i];
	}
}
```

Thanks!

**Reply by pacmano1** [reply]
OBX.4.1 is not '12345' in your example so can you please fix your before and after and post a valid HL7 message (it is missing the hl7 version for example)

**Reply by pacmano1** [reply]
Also post a clear ask.  

"I want to delete an ORC segment and any subsequent segments after _that_ ORC up the next ORC if any OBX after that ORC (and before the next ORC) has <somevalue> in OBX.4.1".    or whatever it actually is.

**Comment by Electro666** [comment]
Yes, thank-you for assisting with what I'm trying to accomplish.  That is exactly what I should be asking. 

I want to delete an ORC segment and any subsequent segments after that ORC up the next ORC if any OBR after that ORC (and before the 
next ORC) has OBR.4.1 = '12345'.

**Reply by pacmano1** [reply]
Fix your sample message above so it matches your use case.  You have 123459999 in it.   Fix the MSH segment.   

Make it easier for people to help you by posting a thorough example.

**Comment by Electro666** [comment]
My example looks good.  You're referencing OBX.3.1 which is not my use case.  I indeed shows '12345' in one of the OBR.4.1 fields.

**Reply by pacmano1** [reply]
Then fix your MSH segment at least.   Simply so the Message Tree in Mirth parses corectly.

```MSH|^~\&|TEST|198|STUFF|CC|20230816095352|11456|ORU^R01|16665|T|2.4```

**Comment by pacmano1** [comment, marked as answer]
Anyway, using @tonygermano code:

```JS
// delete javascript array of segments
function deleteSegList(list) {
    list.forEach(seg => {
        delete seg[0]
    })
}

// delete ORC segment and all consecutive segments until the next ORC or end of message
function deleteOrderObservation(orc) {
    var allSegsUntilNextObr = getSegmentsAfter(msg, orc, /.*/)
    deleteSegList(allSegsUntilNextObr)
    delete orc[0]
}

// Iterate over each ORC
for each(orc in msg.ORC) {
    var obr_segments = getSegmentsAfter(msg, orc, 'OBR', false, 'ORC'). // collect OBRs after ORC
    for each(obr in obr_segments) {
        if (obr['OBR.4']['OBR.4.1'].toString() == '12345') {
            deleteOrderObservation(orc)
            break
        }
    }
}
```

Requires this code tempalte:  https://github.com/nextgenhealthcare/connect-examples/tree/master/Code%20Templates/Get%20Segments%20After%20a%20Particular%20Segment
